[float8 moe training] Add TP support

danielvegamyhre opened the pull request:
Note: this should be merged AFTER this bug fix: https://github.com/pytorch/ao/pull/2451 I will rebase and retest all of this once that's merged.

## Summary
- Add TP support for routed experts and shared expert.
    - Make target dim of scale squeze() ops explicit to handle both 2D and 3D "A" tensors (routed experts case has 2D "A", shared expert has 3D "A").
    - Make `offs` optional to handle shared_expert case where num_experts=1 (scaled grouped GEMM only processing 1 expert)
- Add debug logging

## Test plan
- Added integration test using torchtitan llama4 TP implementation. Test cases for (1) routed experts, and  (2) routed experts + shared expert.
- Manual testing with torchtitan llama4 debug model with TP=2, targeting routed experts AND shared experts works ([logs](https://www.internalfb.com/phabricator/paste/view/P1849252048)). 
- Manual testing with torch titan llama4 debug model with FSDP=2 + TP=2 confirms this 2D parallelism is working for routed experts ([logs](https://www.internalfb.com/phabricator/paste/view/P1848969337))

## Limitations
- 2D parallel witih FDSP+TP for **shared experts** is not yet supported yet (see [comment](https://github.com/pytorch/ao/pull/2425#issuecomment-2997952768)) below). Need to debug this, which I will do in a subsequent PR.

danielvegamyhre commented:
Error with TP on shared expert:

```
+ NGPU=4
+ export LOG_RANK=0
+ LOG_RANK=0
+ CONFIG_FILE=./torchtitan/experiments/llama4/train_configs/debug_model.toml
+ overrides=
+ '[' 5 -ne 0 ']'
+ overrides='--training.steps=100 --model.converters=float8 --float8.recipe_name=rowwise --float8.moe_fqns_prototype=experts,shared_expert --parallelism.tensor_parallel_degree=2'
+ TORCHFT_LIGHTHOUSE=http://localhost:29510
+ PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
+ TORCHFT_LIGHTHOUSE=http://localhost:29510
+ torchrun --nproc_per_node=4 --rdzv_backend c10d --rdzv_endpoint=localhost:0 --local-ranks-filter 0 --role rank --tee 3 -m torchtitan.train --job.config_file ./torchtitan/experiments/llama4/train_configs/debug_model.toml --training.steps=100 --model.converters=float8 --float8.recipe_name=rowwise --float8.moe_fqns_prototype=experts,shared_expert --parallelism.tensor_parallel_degree=2
W0623 13:13:28.914000 2251781 site-packages/torch/distributed/run.py:774] 
W0623 13:13:28.914000 2251781 site-packages/torch/distributed/run.py:774] *****************************************
W0623 13:13:28.914000 2251781 site-packages/torch/distributed/run.py:774] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
W0623 13:13:28.914000 2251781 site-packages/torch/distributed/run.py:774] *****************************************
[rank0]:[titan] 2025-06-23 13:13:35,212 - root - INFO - Starting job: Llama 4 debug training
[rank0]:[titan] 2025-06-23 13:13:36,546 - root - WARNING - ENV[TORCH_NCCL_ASYNC_ERROR_HANDLING] = 1 will be overridden to 3 based on job config
[rank0]:[titan] 2025-06-23 13:13:36,548 - root - INFO - Building 2-D device mesh with ['dp_shard', 'tp'], [2, 2]
[rank0]:[titan] 2025-06-23 13:13:36,551 - root - INFO - [GC] Initial GC collection. 0.00 seconds.
[rank0]:[rank0]:[W623 13:13:40.681077397 Utils.hpp:137] Warning: Environment variable TORCH_NCCL_TRACE_BUFFER_SIZE is deprecated; use TORCH_FR_BUFFER_SIZE instead (function operator())
[rank0]:[titan] 2025-06-23 13:13:45,823 - root - INFO - TikTokenizer built: #words 2256, BOS ID 2000, EOS ID 2001
[rank0]:[titan] 2025-06-23 13:13:45,823 - root - INFO - Preparing c4_test dataset from tests/assets/c4_test
[rank0]:[titan] 2025-06-23 13:13:46,016 - root - INFO - Building llama4 debugmodel with TransformerModelArgs(_enforced='This field is used to enforce all fields have defaults.', dim=256, n_layers=6, n_heads=16, n_kv_heads=None, vocab_size=2256, multiple_of=256, ffn_dim_multiplier=None, norm_eps=1e-05, rope_theta=500000, max_seq_len=2048, depth_init=True, use_flex_attn=False, attn_mask_type='causal', eos_id=2001, every_n_layers_nope=None, fixed_attn_block_size=8192, moe_enabled=True, num_experts=8, use_shared_expert=True, auto_scale_hidden_dim=True, interleave_moe_layer_step=2, top_k=1, use_grouped_mm=True, load_balance_coeff=0.001)
[rank0]:[titan] 2025-06-23 13:13:46,129 - root - INFO - Float8 training active with recipe rowwise
[rank0]:[titan] 2025-06-23 13:13:46,132 - root - INFO - Swapped to Float8Linear layers with enable_fsdp_float8_all_gather=False
[rank0]:[titan] 2025-06-23 13:13:46,142 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w1 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,143 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w2 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,143 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w3 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,143 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w1 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,143 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w2 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,143 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w3 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,144 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w1 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,144 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w2 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,144 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w3 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,144 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w1 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,144 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w2 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,144 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w3 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,145 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w1 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,145 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w2 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,146 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w3 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,146 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w1 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,146 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w2 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,146 - torchao.prototype.moe_training.conversion_utils - INFO - Swapped .w3 to ScaledGroupedMMTensor
[rank0]:[titan] 2025-06-23 13:13:46,146 - root - INFO - Converted MoE to float8
[rank0]:[titan] 2025-06-23 13:13:46,153 - root - INFO - CUDA capacity: NVIDIA H100 with 95.00GiB memory
[rank0]:[titan] 2025-06-23 13:13:46,239 - root - INFO - Total parameter count: dense 4,500,736, sparse 10,622,976, active 6,866,176
[rank0]:[titan] 2025-06-23 13:13:46,239 - root - INFO - Model llama4 debugmodel size: 15,123,712 total parameters
[rank0]:[titan] 2025-06-23 13:13:46,258 - root - INFO - Applied Tensor Parallelism to the model
[rank0]:[titan] 2025-06-23 13:13:46,303 - root - INFO - Applied FSDP to the model
[rank0]:/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py:1355: UserWarning: Casting complex values to real discards the imaginary part (Triggered internally at /pytorch/aten/src/ATen/native/Copy.cpp:307.)
[rank0]:  return t.to(
[rank0]:[titan] 2025-06-23 13:13:46,775 - root - INFO - Peak FLOPS used for computing MFU: 9.890e+14
[rank0]:[titan] 2025-06-23 13:13:46,776 - root - INFO - CUDA memory usage for model: 0.02GiB(0.02%)
[rank0]:[titan] 2025-06-23 13:13:46,776 - root - INFO - Trainer is initialized with local batch size 8, global batch size 16, sequence length 2048, total steps 100 (warmup 2).
[rank0]:[titan] 2025-06-23 13:13:46,777 - root - INFO - Training starts at step 1.
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721] failed while attempting to run meta for aten._grouped_mm.default
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721] Traceback (most recent call last):
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_subclasses/fake_tensor.py", line 2717, in _dispatch_impl
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     r = func(*args, **kwargs)
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_ops.py", line 829, in __call__
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     return self._op(*args, **kwargs)
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]            ~~~~~~~~^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_prims_common/wrappers.py", line 309, in _fn
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     result = fn(*args, **kwargs)
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_meta_registrations.py", line 7677, in grouped_mm
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     return _meta_grouped_mm_common(
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]         mat_a,
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     ...<6 lines>...
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]         out_dtype=out_dtype,
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     )
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_meta_registrations.py", line 7527, in _meta_grouped_mm_common
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     torch._check(
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     ~~~~~~~~~~~~^
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]         mat_a.dtype == torch.bfloat16 and mat_b.dtype == torch.bfloat16,
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]         lambda: f"Expected inputs of BF16 type but got mat_a.dtype={mat_a.dtype} and mat_b.dtype={mat_b.dtype}.",
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     )
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     ^
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/__init__.py", line 1702, in _check
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     _check_with(RuntimeError, cond, message)
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/__init__.py", line 1684, in _check_with
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721]     raise error_type(message_evaluated)
[rank0]:[rank0]:E0623 13:13:49.827000 2252339 site-packages/torch/_subclasses/fake_tensor.py:2721] RuntimeError: Expected inputs of BF16 type but got mat_a.dtype=torch.bfloat16 and mat_b.dtype=torch.float32.
[rank0]:[titan] 2025-06-23 13:13:50,799 - root - INFO - Process group destroyed.
[rank0]:[rank0]: Traceback (most recent call last):
[rank0]:[rank0]:   File "<frozen runpy>", line 198, in _run_module_as_main
[rank0]:[rank0]:   File "<frozen runpy>", line 88, in _run_code
[rank0]:[rank0]:   File "/home/danvm/torchtitan/torchtitan/train.py", line 495, in <module>
[rank0]:[rank0]:     trainer.train()
[rank0]:[rank0]:     ~~~~~~~~~~~~~^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 357, in wrapper
[rank0]:[rank0]:     return f(*args, **kwargs)
[rank0]:[rank0]:   File "/home/danvm/torchtitan/torchtitan/train.py", line 437, in train
[rank0]:[rank0]:     self.train_step(inputs, labels)
[rank0]:[rank0]:     ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/torchtitan/torchtitan/train.py", line 370, in train_step
[rank0]:[rank0]:     pred = model_parts[0](inputs)
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
[rank0]:[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:[rank0]:            ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1879, in _call_impl
[rank0]:[rank0]:     return inner()
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1827, in inner
[rank0]:[rank0]:     result = forward_call(*args, **kwargs)
[rank0]:[rank0]:   File "/home/danvm/torchtitan/torchtitan/experiments/llama4/model/model.py", line 471, in forward
[rank0]:[rank0]:     h = layer(h, self.freqs_cis)
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
[rank0]:[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:[rank0]:            ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1879, in _call_impl
[rank0]:[rank0]:     return inner()
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1827, in inner
[rank0]:[rank0]:     result = forward_call(*args, **kwargs)
[rank0]:[rank0]:   File "/home/danvm/torchtitan/torchtitan/experiments/llama4/model/model.py", line 339, in forward
[rank0]:[rank0]:     out = h + self.moe(self.ffn_norm(h))
[rank0]:[rank0]:               ~~~~~~~~^^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
[rank0]:[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:[rank0]:            ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1879, in _call_impl
[rank0]:[rank0]:     return inner()
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1827, in inner
[rank0]:[rank0]:     result = forward_call(*args, **kwargs)
[rank0]:[rank0]:   File "/home/danvm/torchtitan/torchtitan/experiments/llama4/model/moe.py", line 318, in forward
[rank0]:[rank0]:     routed_output = self.experts(routed_input, num_local_tokens_per_expert)
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
[rank0]:[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:[rank0]:            ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1879, in _call_impl
[rank0]:[rank0]:     return inner()
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1827, in inner
[rank0]:[rank0]:     result = forward_call(*args, **kwargs)
[rank0]:[rank0]:   File "/home/danvm/torchtitan/torchtitan/experiments/llama4/model/moe.py", line 91, in forward
[rank0]:[rank0]:     h = F.silu(torch._grouped_mm(x, self.w1, offs=offsets))
[rank0]:[rank0]:                ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_compile.py", line 53, in inner
[rank0]:[rank0]:     return disable_fn(*args, **kwargs)
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_dynamo/eval_frame.py", line 896, in _fn
[rank0]:[rank0]:     return fn(*args, **kwargs)
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/tensor/_api.py", line 350, in __torch_dispatch__
[rank0]:[rank0]:     return DTensor._op_dispatcher.dispatch(
[rank0]:[rank0]:            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
[rank0]:[rank0]:         func,
[rank0]:[rank0]:         ^^^^^
[rank0]:[rank0]:         args,
[rank0]:[rank0]:         ^^^^^
[rank0]:[rank0]:         kwargs or {},
[rank0]:[rank0]:         ^^^^^^^^^^^^^
[rank0]:[rank0]:     )
[rank0]:[rank0]:     ^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/tensor/_dispatch.py", line 154, in dispatch
[rank0]:[rank0]:     self.sharding_propagator.propagate(op_info)
[rank0]:[rank0]:     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/tensor/_sharding_prop.py", line 266, in propagate
[rank0]:[rank0]:     OutputSharding, self.propagate_op_sharding(op_info.schema)
[rank0]:[rank0]:                     ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/tensor/_sharding_prop.py", line 45, in __call__
[rank0]:[rank0]:     return self.cache(*args, **kwargs)
[rank0]:[rank0]:            ~~~~~~~~~~^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/tensor/_sharding_prop.py", line 279, in propagate_op_sharding_non_cached
[rank0]:[rank0]:     out_tensor_meta = self._propagate_tensor_meta_non_cached(op_schema)
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/tensor/_sharding_prop.py", line 126, in _propagate_tensor_meta_non_cached
[rank0]:[rank0]:     fake_out = op_schema.op(*fake_args, **fake_kwargs)
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_ops.py", line 829, in __call__
[rank0]:[rank0]:     return self._op(*args, **kwargs)
[rank0]:[rank0]:            ~~~~~~~~^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/utils/_stats.py", line 28, in wrapper
[rank0]:[rank0]:     return fn(*args, **kwargs)
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_subclasses/fake_tensor.py", line 1352, in __torch_dispatch__
[rank0]:[rank0]:     return self.dispatch(func, types, args, kwargs)
[rank0]:[rank0]:            ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_subclasses/fake_tensor.py", line 2058, in dispatch
[rank0]:[rank0]:     return self._cached_dispatch_impl(func, types, args, kwargs)
[rank0]:[rank0]:            ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_subclasses/fake_tensor.py", line 1487, in _cached_dispatch_impl
[rank0]:[rank0]:     output = self._dispatch_impl(func, types, args, kwargs)
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_subclasses/fake_tensor.py", line 2717, in _dispatch_impl
[rank0]:[rank0]:     r = func(*args, **kwargs)
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_ops.py", line 829, in __call__
[rank0]:[rank0]:     return self._op(*args, **kwargs)
[rank0]:[rank0]:            ~~~~~~~~^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_prims_common/wrappers.py", line 309, in _fn
[rank0]:[rank0]:     result = fn(*args, **kwargs)
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_meta_registrations.py", line 7677, in grouped_mm
[rank0]:[rank0]:     return _meta_grouped_mm_common(
[rank0]:[rank0]:         mat_a,
[rank0]:[rank0]:     ...<6 lines>...
[rank0]:[rank0]:         out_dtype=out_dtype,
[rank0]:[rank0]:     )
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_meta_registrations.py", line 7527, in _meta_grouped_mm_common
[rank0]:[rank0]:     torch._check(
[rank0]:[rank0]:     ~~~~~~~~~~~~^
[rank0]:[rank0]:         mat_a.dtype == torch.bfloat16 and mat_b.dtype == torch.bfloat16,
[rank0]:[rank0]:         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:         lambda: f"Expected inputs of BF16 type but got mat_a.dtype={mat_a.dtype} and mat_b.dtype={mat_b.dtype}.",
[rank0]:[rank0]:         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:     )
[rank0]:[rank0]:     ^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/__init__.py", line 1702, in _check
[rank0]:[rank0]:     _check_with(RuntimeError, cond, message)
[rank0]:[rank0]:     ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:[rank0]:   File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/__init__.py", line 1684, in _check_with
[rank0]:[rank0]:     raise error_type(message_evaluated)
[rank0]:[rank0]: RuntimeError: Expected inputs of BF16 type but got mat_a.dtype=torch.bfloat16 and mat_b.dtype=torch.float32.
W0623 13:13:52.843000 2251781 site-packages/torch/distributed/elastic/multiprocessing/api.py:900] Sending process 2252339 closing signal SIGTERM
W0623 13:13:52.843000 2251781 site-packages/torch/distributed/elastic/multiprocessing/api.py:900] Sending process 2252340 closing signal SIGTERM
W0623 13:13:52.843000 2251781 site-packages/torch/distributed/elastic/multiprocessing/api.py:900] Sending process 2252344 closing signal SIGTERM
E0623 13:13:53.570000 2251781 site-packages/torch/distributed/elastic/multiprocessing/api.py:874] failed (exitcode: 1) local_rank: 2 (pid: 2252341) of binary: /home/danvm/.conda/envs/torchtitan/bin/python
E0623 13:13:53.573000 2251781 site-packages/torch/distributed/elastic/multiprocessing/errors/error_handler.py:141] no error file defined for parent, to copy child error file (/tmp/torchelastic_rrfserz5/none_n65wzddi/attempt_0/2/error.json)
[rank0]:NCCL version 2.27.3+cuda12.9
[rank0]:X dtype: torch.bfloat16
[rank0]:W1 dtype: torch.bfloat16
[rank0]:W1 type: <class 'torch.distributed.tensor.DTensor'>
[rank0]:W1.to_local() type: <class 'torchao.prototype.moe_training.tensor.ScaledGroupedMMTensor'>
[rank0]:W1.to_local() dtype: torch.bfloat16
Traceback (most recent call last):
  File "/home/danvm/.conda/envs/torchtitan/bin/torchrun", line 8, in <module>
    sys.exit(main())
             ~~~~^^
  File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 357, in wrapper
    return f(*args, **kwargs)
  File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/run.py", line 901, in main
    run(args)
    ~~~^^^^^^
  File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/run.py", line 892, in run
    elastic_launch(
    ~~~~~~~~~~~~~~~
        config=config,
        ~~~~~~~~~~~~~~
        entrypoint=cmd,
        ~~~~~~~~~~~~~~~
    )(*cmd_args)
    ~^^^^^^^^^^^
  File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/launcher/api.py", line 143, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/launcher/api.py", line 277, in launch_agent
    raise ChildFailedError(
    ...<2 lines>...
    )
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
torchtitan.train FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2025-06-23_13:13:49
  host      : devgpu007.eag6.facebook.com
  rank      : 2 (local_rank: 2)
  exitcode  : 1 (pid: 2252341)
  error_file: /tmp/torchelastic_rrfserz5/none_n65wzddi/attempt_0/2/error.json
  traceback : Traceback (most recent call last):
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 357, in wrapper
      return f(*args, **kwargs)
    File "/home/danvm/torchtitan/torchtitan/train.py", line 437, in train
      self.train_step(inputs, labels)
      ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
    File "/home/danvm/torchtitan/torchtitan/train.py", line 370, in train_step
      pred = model_parts[0](inputs)
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
      return self._call_impl(*args, **kwargs)
             ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1879, in _call_impl
      return inner()
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1827, in inner
      result = forward_call(*args, **kwargs)
    File "/home/danvm/torchtitan/torchtitan/experiments/llama4/model/model.py", line 471, in forward
      h = layer(h, self.freqs_cis)
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
      return self._call_impl(*args, **kwargs)
             ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1879, in _call_impl
      return inner()
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/nn/modules/module.py", line 1806, in inner
      args_kwargs_result = hook(self, args, kwargs)  # type: ignore[misc]
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_state.py", line 62, in fsdp_hook_wrapper
      return torch._dynamo.disable(
             ~~~~~~~~~~~~~~~~~~~~~~
      ...<2 lines>...
          reason="skipping FSDP hooks since torch._dynamo.config.skip_fsdp_hooks is set",
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      )(*args, **kwargs)
      ~^^^^^^^^^^^^^^^^^
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_dynamo/eval_frame.py", line 896, in _fn
      return fn(*args, **kwargs)
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_state.py", line 248, in _pre_forward
      args, kwargs = self._fsdp_param_group.pre_forward(module, args, kwargs)
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py", line 352, in pre_forward
      self.wait_for_unshard()
      ~~~~~~~~~~~~~~~~~~~~~^^
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py", line 311, in wait_for_unshard
      fsdp_param.init_unsharded_param()
      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/distributed/fsdp/_fully_shard/_fsdp_param.py", line 504, in init_unsharded_param
      unsharded_param = torch.as_strided(
          unsharded_tensor,
      ...<2 lines>...
          storage_offset=0,
      )
    File "/home/danvm/ao/torchao/prototype/moe_training/tensor.py", line 93, in __torch_function__
      return func(*args, **kwargs)
    File "/home/danvm/ao/torchao/prototype/moe_training/tensor.py", line 108, in __torch_dispatch__
      out = func(*args, **kwargs)
    File "/home/danvm/.conda/envs/torchtitan/lib/python3.13/site-packages/torch/_ops.py", line 829, in __call__
      return self._op(*args, **kwargs)
             ~~~~~~~~^^^^^^^^^^^^^^^^^
  RuntimeError: setStorage: sizes [1, 256, 256], strides [65536, 256, 1], storage offset 0, and itemsize 2 requiring a storage size of 131072 are out of bounds for storage of size 0

```

drisspg commented on test/prototype/moe_training/test_tp.py:
```diff
@@ -0,0 +1,219 @@
+
+# this test requires torchtitan
+try:
+    from torchtitan.experiments.llama4.infra.parallelize import apply_moe_tp
```
we should add this test to test_float8 -> https://github.com/pytorch/ao/blob/994a4ba6c869854fcaa6ca7e118fcbd75e6c28cc/.github/workflows/float8_test.yml#L46

drisspg commented on test/prototype/moe_training/test_tp.py:
```diff
@@ -0,0 +1,219 @@
+    return device_mesh
+
+
+def apply_moe_tp(
```
this is always specific to module structure e.g. the fqn's right?

danielvegamyhre commented:
Closing in favor of https://github.com/pytorch/ao/pull/2473 but @drisspg  I addressed your comments in the new PR. The new PR is not ready for review yet, still some work to do to support the shared expert for TP. Will tag you when it's ready
